### ISSUE TITLE
Example for $? is not valid

### ISSUE THREAD

**Issue by VShrubowich** [MicrosoftDocs/PowerShell-Docs#9877]
### Prerequisites

- [X] **Existing Issue:** Search the existing issues for this repository. If there is an issue that fits your needs do not file a new one. Subscribe, react, or comment on that issue instead.
- [X] **Descriptive Title:** Write the title for this issue as a short synopsis. If possible, provide context. For example, "Typo in `Get-Foo` cmdlet" instead of "Typo."
- [X] **Verify Version:** If there is a mismatch between documentation and the behavior on your system, ensure that the version you are using is the same as the documentation. Check this box if they match or the issue you are reporting is not version specific.

### Links

[https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_automatic_variables#section-1](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_automatic_variables#section-1)

### Summary

The example for `$?` is invalid because the `$?` in the function resets `$?` to **True** after successfully displaying **False**. Therefore, even if the `Write-Error` cmdlet did set `$?` to **False** for a function calling it, the `$?` in the function would reset it back to **True** and calling the function would still return **True**.

### Details

The following shows that the `$?` in the function resets `$?` to **True**:

```powershell
function Test-WriteError
{
    Write-Error "Bad"
    $? # $false
    $? # $true
}

Test-WriteError
```

```output
Test-WriteError: Bad
False
True
```

### Suggested Fix

The following is one way to fix it:

```powershell
function Test-WriteError
{
    Write-Error "Bad"
    $? # $false
    Write-Error "Bad again" # Writing error again since $? above succeeded, resetting $?
}

Test-WriteError
$? # $true
```

**Comment by michaeltlombardi** [comment]
Hi, @VShrubowich, thanks for filing this issue!

I've updated the example in #9882 for clarity.

The example is actually valid - the point is displaying that the $? always represents the status for the last statement. The original example was showing this but in a less-than obvious way by simply emitting the value for `$?` in the function (after the error set it to `$false`) and after the function call (which sets it to `$true`).

**Comment by VShrubowich** [comment]
@michaeltlombardi, I may be misunderstanding the point, but I still believe it's not valid, even as rewritten, since emitting the value of `$?` or the string ``"The `$? variable is: $?"`` both succeed, setting `$?` to **True** once again.  When I run the following:

```powershell
function Test-WriteError
{
    Write-Error "Bad"
    "The `$? variable is: $?"
    "Now the `$? variable is: $?"
}

Test-WriteError
```

I get the following:

```Output
Test-WriteError: Bad
The $? variable is: False
Now the $? variable is: True
```

### A Suggested Fix

The Write-Error cmdlet always sets `$?` to **False** immediately after it's executed, but won't set `$?` to **False** for a function calling it.

For example, checking `$?` within the function shows that `$?` is **False**:

```powershell
function Test-WriteError
{
    Write-Error "Bad"
    "Checking `$? within the function is: $?"
}

Test-WriteError
```

```Output
Test-WriteError: Bad
Checking $? within the function is: False
```

However, checking `$?` outside of the function shows that `$?` is now **True**:

```powershell
function Test-WriteError
{
    Write-Error "Bad"
}

Test-WriteError; "Checking `$? outside of the function is: $?"
```

```Output
Test-WriteError: Bad
Checking $? outside of the function is: True
```

**Comment by michaeltlombardi** [comment]
> I may be misunderstanding the point, but I still believe it's not valid, even as rewritten, since emitting the value of `$?` or the string `"The `$? variable is: $?"` both succeed, setting `$?` to **True** once again.

That is precisely the point of the example. After `$?` is set after each statement.

It's showing that while `$?` is **False** _in the function after an error_, it's **True** afterward _because_ the emitting statement succeeded.

**Comment by VShrubowich** [comment]
I figured out the history of why the example, and the lines immediately before and after the example, were added to the documentation.  It stems from PowerShell issue #[3629](https://github.com/PowerShell/PowerShell/issues/3629), **Write-Error in a function doesn't set automatic success variable $? to $False (doesn't set the execution status to indicate a non-terminating error) in the caller's scope**.  This led to PowerShell-Docs issue #[3102](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/3102), **Docs needed for 'Write-Error in a function doesn't set automatic success variable $? to $False (doesn't set the execution status to indicate a non-terminating error) in the caller's scope'**,

You mentioned the following:

> It's showing that while `$?` is **False** _in the function after an error_, it's **True** afterward _because_ the emitting statement succeeded.

By _emitting statement_, I believe you mean ``"The `$? variable is: $?"`` within the function.  However, the example is supposed to show that `$?` is not set to **False** _in the caller's scope_ when `Write-Error` is used _in a function_.  Including ``"The `$? variable is: $?"`` in the example is misleading since that will reset `$?` to **True**, resulting in `$?` being **True** even if `Write-Error` was able to set `$?` to **False** in the caller's scope.

The PowerShell [issue](https://github.com/PowerShell/PowerShell/issues/3629#issue-223639068) was opened with the following being reported:

> `Write-Error` is used to signal a non-terminating error, yet `$?` (the execution-status automatic variable) isn't set to `$False` in the _caller's_ scope when it is called from an (advanced) function.
> 
> By contrast, if you use `$PSCmdlet.WriteError` from an advanced function, it is.
> 
> ### Steps to reproduce
> 
> ```powershell
> function foo { Write-Error 'oh no' }; foo; $?
> function foo2 { [cmdletbinding()] param() $PSCmdlet.WriteError((New-Object System.Management.Automation.ErrorRecord "oh no", $null, 'NotSpecified', $null)) }; foo2; $?
> ```
> 
> ### Expected behavior
> 
> ```powershell
> # (error details omitted)
> $False
> # (error details omitted)
> $False
> ```
> 
> ### Actual behavior
> 
> ```powershell
> # (error details omitted)
> $True
> # (error details omitted)
> $False
> ```
> 

---

To be faithful to the original issue, I suggest replacing the following:

> The `Write-Error` cmdlet always sets `$?` to **False** immediately after it's executed, but won't set `$?` to **False** for a function calling it:
> 
> `(PowerShell code and Output)`
> 
> For the latter purpose, `$PSCmdlet.WriteError()` should be used instead.

with the following Suggested Fixes.

## Suggested Fix (PowerShell 7.x).

The `Write-Error` cmdlet always sets `$?` to **False** immediately after it's executed, but won't set `$?` to **False** in the caller's scope when used in a function:

```powershell
function Test-WriteError
{
    Write-Error "Bad"
}

# Calling the function
Test-WriteError
"The `$? variable is: $?"
```

```Output
Test-WriteError:
Line |
   6 |  Test-WriteError
     |  ~~~~~~~~~~~~~~~
     | Bad
The $? variable is: True
```

To have `$?` set to **False** in the caller's scope, change the example above to be an [advanced function](about_functions_advanced) using $PSCmdlet.WriteError():

```powershell
function Test-WriteError
{
    [CmdletBinding()]
    Param()
    $PSCmdlet.WriteError((New-Object System.Management.Automation.ErrorRecord "Bad", $null, 'NotSpecified', $null))
}

# Calling the advanced function
Test-WriteError
"The `$? variable is: $?"
```

```Output
Test-WriteError: 
Line |
   9 |  Test-WriteError
     |  ~~~~~~~~~~~~~~~
     | Bad
The $? variable is: False
```

## Suggested Fix (PowerShell 5.1).

The `Write-Error` cmdlet always sets `$?` to **False** immediately after it's executed, but won't set `$?` to **False** in the caller's scope when used in a function:

```powershell
function Test-WriteError
{
    Write-Error "Bad"
}

# Calling the function
Test-WriteError
"The `$? variable is: $?"
```

```Output
Test-WriteError : Bad
At line:6 char:1
+ Test-WriteError
+ ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Test-WriteError

The $? variable is: True
```

To have `$?` set to **False** in the caller's scope, change the example above to be an [advanced function](about_functions_advanced) using $PSCmdlet.WriteError():

```powershell
function Test-WriteError
{
    [CmdletBinding()]
    Param()
    $PSCmdlet.WriteError((New-Object System.Management.Automation.ErrorRecord "Bad", $null, 'NotSpecified', $null))
}

# Calling the advanced function
Test-WriteError
"The `$? variable is: $?"
```

```Output
Test-WriteError : Bad
At line:9 char:1
+ Test-WriteError
+ ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Test-WriteError], Exception
    + FullyQualifiedErrorId : Test-WriteError

The $? variable is: False
```

---

# Alternative Suggestion

The above is faithful to the original issue, However, if you would prefer to show the _differences_ between how `Write-Error` and `$PSCmdlet.WriteError()` affect `$?`, then everything can be left as is except editing "For the latter purpose, `$PSCmdlet.WriteError()` should be used instead." and adding an example after it:

## PowerShell 7.x Example

For the latter purpose, an [advanced function](about_functions_advanced) with `$PSCmdlet.WriteError()` should be used instead:

```powershell
function Test-WriteError
{
    [CmdletBinding()]
    Param()
    $PSCmdlet.WriteError((New-Object System.Management.Automation.ErrorRecord "Bad", $null, 'NotSpecified', $null))
    "The `$? variable is: $?"
}

Test-WriteError
"Now the `$? variable is: $?"
```

```Output
Test-WriteError: 
Line |
  10 |  Test-WriteError
     |  ~~~~~~~~~~~~~~~
     | Bad
The $? variable is: True
Now the $? variable is: False
```

## PowerShell 5.1 Example

For the latter purpose, an [advanced function](about_functions_advanced) with `$PSCmdlet.WriteError()` should be used instead:

```powershell
function Test-WriteError
{
    [CmdletBinding()]
    Param()
    $PSCmdlet.WriteError((New-Object System.Management.Automation.ErrorRecord "Bad", $null, 'NotSpecified', $null))
    "The `$? variable is: $?"
}

Test-WriteError
"Now the `$? variable is: $?"
```

```Output
Test-WriteError : Bad
At line:10 char:1
+ Test-WriteError
+ ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Test-WriteError], Exception
    + FullyQualifiedErrorId : Test-WriteError

The $? variable is: True
Now the $? variable is: False
```
